test/common/upstream/health_checker_impl_test.cc

```diff
@@ -614,6 +614,56 @@ TEST_F(HttpHealthCheckerImplTest, RemoteCloseBetweenChecks) {
   EXPECT_TRUE(cluster_->hosts_[0]->healthy());
 }
 
+TEST_F(HttpHealthCheckerImplTest, StreamReachesWatermarkDuringCheck) {
+  setupNoServiceValidationHC();
+  EXPECT_CALL(*this, onHostStatus(_, false));
+
+  cluster_->hosts_ = {makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, enableTimer(_));
+  health_checker_->start();
+
+  EXPECT_CALL(*test_sessions_[0]->interval_timer_, enableTimer(_));
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, disableTimer());
+
+  for (auto* callback : test_sessions_[0]->request_encoder_.stream_.callbacks_) {
```

Comment:
nit: I might add `runHighWatermarkCallbacks()` and `runLowWatermarkCallbacks()` here: https://github.com/envoyproxy/envoy/blob/master/test/mocks/http/mocks.h#L153. We may want to use this elsewhere in the future and we can put the loop inside the mock.